src/LinearSolverTypes.f90

```diff
@@ -1431,17 +1427,53 @@ SUBROUTINE getResidual_LinearSolverType_Iterative(solver,resid)
     IF(resid%n == solver%b%n) THEN
 #ifdef HAVE_MKL
       !not yet implemented
-      SELECTTYPE(b => solver%b); TYPE IS(RealVectorType)
-        resid%b=-b%b
+      SELECTTYPE(b => solver%b)
+      TYPE IS(RealVectorType)
+        SELECTTYPE(resid)
+        TYPE IS(RealVectorType)
+          resid%b=-b%b
+          CALL BLAS_matvec(THISMATRIX=solver%A,X=solver%X,Y=resid,ALPHA=1.0_SRK,BETA=1.0_SRK)
+        CLASS DEFAULT
+          CALL eLinearSolverType%raiseError('Incorrect call to '// &
+              modName//'::getResidual'//' - Type mismatch between getResidual argument and solver datatypes')
+        ENDSELECT
+#ifdef HAVE_MPI
+      TYPE IS(NativeDistributedVectorType)
+        SELECTTYPE(resid)
+        TYPE IS(NativeDistributedVectorType)
+          resid%b=-b%b
+          CALL BLAS_matvec(THISMATRIX=solver%A,X=solver%X,Y=resid,ALPHA=1.0_SRK,BETA=1.0_SRK)
+        CLASS DEFAULT
+          CALL eLinearSolverType%raiseError('Incorrect call to '// &
+              modName//'::getResidual'//' - Type mismatch between getResidual argument and solver datatypes')
+        ENDSELECT
```

Comment:
Would it be better to put the serial portion of the code into an `#else` block?

Reply:
This code's been refactored; getresidual has been moved to 2 routines, one for the residual norm and one for the residual vector (and iteration count).
This makes it a more common and reliable interface overall in the LinearSolverTypes framework which alleviates some headaches when integrating with MPACT.